book-content/chapters/02-ide-superpowers.md

```diff
@@ -0,0 +1,579 @@
+
+a.toString();
+```
+
+TypeScript tells us that there is a problem with `a`. This tells us where the problem is, but it doesn't necessarily tell us what the problem is. In this case, we need to stop and think about why we can't call `toString()` on `null`. If we do, it will throw an error at runtime.
+
+```
+Uncaught TypeError: Cannot read properties of null (reading 'toString').
+```
+
+Here, TypeScript is telling us that an error might happen even without us needing to check it. Very handy.
+
+### Warnings About Non-Runtime Errors
+
+Not everything TypeScript warns us about will actually fail at runtime.
+
+Take a look at this example where we're assigning a property to an empty object:
+
+```ts twoslash
+// @errors: 2339
+const obj = {};
+
+const result = obj.foo;
+```
+
+TypeScript draws a red squiggly line below `foo`. But if we think about it, this code won't actually cause an error at runtime. We're trying to assign a property that doesn't exist in this object: `foo`. This won't error, it will just mean that result is undefined.
+
+It may seem strange that TypeScript would warn us about something that won't cause an error, but it's actually a good thing. If TypeScript didn't warn us about this, it would be saying that we can access any property on any object at any time. Over the course of an entire application, this could result in quite a few bugs.
+
+It's best to think of TypeScript's rules as opinionated. They are a collection of helpful tips that will make your application safer as a whole.
+
+### Warnings Close to the Source of the Problem
+
+TypeScript will try to give you warnings as close to the source of the problem as possible.
+
+Let's take a look at an example.
+
+```ts twoslash
+// @errors: 2561
+type Album = {
+  artist: string;
+  title: string;
+  year: number;
+};
+
+const album: Album = {
+  artsist: "Television",
+  title: "Marquee Moon",
+  year: 1977,
+};
+```
+
+We define an 'Album' type - an object with three properties. Then, we say that const album needs to be of that type via `const album: Album`. Don't worry if you don't understand all the syntax yet - we'll cover it all later.
+
+Can you see the problem? There's a typo of the `artist` property when creating an album. That's because we've said that the `album` variable needs to be of type `Album`, but we've misspelled `artist` as `artsist`. TypeScript is telling us that we've made a mistake, and even suggests the correct spelling.
+
+### Dealing With Multi-Line Errors
+
+However, sometimes TypeScript will give you an error in a more unhelpful spot.
+
+In this example, we have a function called `logUserJobTitle` that logs the job title of a user:
+
+```typescript
+const logUserJobTitle = (user: {
+  job: {
+    title: string;
+  };
+}) => {
+  console.log(user.job.title);
+};
+```
+
+Don't worry about the syntax for now - we'll cover it later. The important thing is that `logUserJobTitle` takes a user object with a `job` property that has a `title` property.
```

Comment:
"Don't worry about the syntax for now - we'll cover it later." is already mentioned like 3 paragraphs above. I'd skip it here as it sounds a little bit patronizing when repeated